pkg/operator/applier/kubectl.go

```diff
@@ -132,6 +136,138 @@ commonAnnotations:
 	return stdout, stderr, errors.Wrap(err, "failed to run kubectl apply")
 }
 
+// ApplyCreateOrPatch attempts to run a `kubectl apply` on the yaml document. If it fails
+func (c *Kubectl) ApplyCreateOrPatch(targetNamespace string, slug string, yamlDoc []byte, dryRun bool, wait bool, annotateSlug bool) ([]byte, []byte, error) {
+
+	stdout, stderr, err := c.Apply(targetNamespace, slug, yamlDoc, dryRun, wait, annotateSlug)
+	if !strings.Contains(string(stderr), "metadata.annotations: Too long") || err == nil {
+		return stdout, stderr, errors.Wrap(err, "failed attempted kubectl apply")
+	}
+	log.Println("Failed to apply document: metadata too long. Trying to 'create' or 'patch' instead")
+
+	docs, err := splitter.SplitYAML(yamlDoc)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "split yaml")
+	}
+
+	var combinedStdout, combinedStderr []byte
+	for name, b := range docs {
+		stdout, stderr, err = c.createOrPatchSingleDoc(name, targetNamespace, slug, b, dryRun, wait, annotateSlug)
+		combinedStdout = append(combinedStdout, stdout...)
+		combinedStderr = append(combinedStderr, stderr...)
+
+		if err != nil {
+			return combinedStdout, combinedStderr, errors.Wrap(err, "failed to run kubectl create")
+		}
+	}
+
+	return combinedStdout, combinedStderr, nil
+}
+
+func (c *Kubectl) createOrPatchSingleDoc(name string, targetNamespace string, slug string, yamlDoc []byte, dryRun bool, wait bool, annotateSlug bool) ([]byte, []byte, error) {
+	tmp, err := ioutil.TempDir("", "")
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create temp directory")
+	}
+	defer os.Remove(tmp)
+
+	docPath := filepath.Join(tmp, "doc.yaml")
+	if err := ioutil.WriteFile(docPath, yamlDoc, 0644); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to write %s", docPath)
+	}
+
+	kustomizationPath := filepath.Join(tmp, "kustomization.yaml")
+	kustomizationYaml := `
+resources:
+- doc.yaml
+`
+	if annotateSlug {
+		kustomizationYaml = fmt.Sprintf(`
+resources:
+- doc.yaml
+
+commonAnnotations:
+  kots.io/app-slug: %q
+`, slug)
+	}
+	if err := ioutil.WriteFile(kustomizationPath, []byte(kustomizationYaml), 0644); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to write %s", kustomizationPath)
+	}
+
+	renderedManifest, err := exec.Command(kustomize.GetKustomizePath(kotsutil.KotsKustomizeVersion), "build", tmp).Output()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to run kustomize build")
+	}
+
+	renderedManifestPath := filepath.Join(tmp, "patch.yaml")
+	if err := ioutil.WriteFile(renderedManifestPath, renderedManifest, 0644); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to write %s", renderedManifestPath)
+	}
+	// Try to create and checkout output to see if it already exists
+	cmd := c.kubectlCreateCommand(renderedManifestPath, targetNamespace, dryRun, wait)
+
+	stdout, stderr, err := Run(cmd)
+	if !strings.Contains(string(stderr), "already exists") || err == nil {
```

Comment:
i wonder if we could check the returned err for something like
```
kuberneteserrors.IsNotFound(err)
```
using the `k8s.io/apimachinery/pkg/api/errors` package

Reply:
That would be better but since we are exec-ing the kubectl binary, I wasn't sure how to get this error type back. Is there any easy way to parse this back into `kuberneteserrrors`?